lib/protocol/protocol.go

```diff
@@ -974,7 +984,9 @@ func (c *rawConnection) internalClose(err error) {
 			<-c.dispatcherLoopStopped
 		}
 
-		c.model.Closed(err)
+		// We don't want to call into the model while holding the
+		// startStopMut, so defer it into the background.
+		go c.model.Closed(err)
```

Comment:
Very optional suggestion: Given how small and simple this function is, we could also unlock before this instead of deferring (moving the locking inside the once, which imo is anyway a bit more consistent).